lib/membrane_tcp/source.ex

```diff
@@ -65,49 +71,61 @@ defmodule Membrane.TCP.Source do
         {:client, _server_address, _server_port_no} -> :client
       end
 
-    {[],
+    actions =
+      case local_socket do
+        %Socket{socket_handle: nil} ->
+          []
+
+        %Socket{socket_handle: handle} ->
+          [notify_parent: {:request_socket_control, handle, self()}]
+      end
+
+    {actions,
      %{
        connection_side: connection_side,
        local_socket: local_socket,
        remote_socket: remote_socket
      }}
   end
 
+  @impl true
+  defdelegate handle_setup(context, state), to: CommonSocketBehaviour
+
   @impl true
   def handle_playing(_ctx, state) do
     {[stream_format: {:output, %RemoteStream{type: :bytestream}}], state}
   end
 
   @impl true
-  def handle_demand(_pad, _size, _unit, _ctx, state) do
-    case Socket.recv(state.local_socket) do
-      {:ok, payload} ->
-        {:ok, {peer_address, peer_port_no}} = :inet.peername(state.local_socket.socket_handle)
-
-        metadata =
-          Map.new()
-          |> Map.put(:tcp_source_address, peer_address)
-          |> Map.put(:tcp_source_port, peer_port_no)
-          |> Map.put(:arrival_ts, Membrane.Time.vm_time())
-
-        {
-          [buffer: {:output, %Buffer{payload: payload, metadata: metadata}}, redemand: :output],
-          state
-        }
-
-      {:error, :timeout} ->
-        {[redemand: :output], state}
-
-      {:error, :closed} ->
-        {[end_of_stream: :output], state}
-
-      {:error, reason} ->
-        raise "TCP Socket receiving error, reason: #{inspect(reason)}"
-    end
+  def handle_demand(_pad, size, _unit, _ctx, state) do
+    :inet.setopts(state.local_socket.socket_handle, active: size)
```

Comment:
This makes sense if only if the demand unit is buffers, you can enforce it in `def_output_pad`